torchaudio/datasets/datasets.py

```diff
@@ -0,0 +1,419 @@
+            self._cache.append(next(self.generator))
+
+    def __getitem__(self, n):
+        self._fill()
+        return self._cache[n]
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        item = self._cache.pop(0)
+        self._fill()
+        return item
+
+
+def download(urls, root_path):
+    """
+    Download each url to root_path.
+
+    Input: url generator, folder inside archive
+    Output: downloaded archive, folder inside archive
+    """
+    for url, folder in urls:
+        torchaudio.datasets.utils.download_url(url, root_path)
+        file = os.path.join(root_path, os.path.basename(url))
+        yield file, folder
+
+
+def extract(files):
+    """
+    Extract each archive to their respective folder.
+
+    Input: (url, folder name inside archive) generator
+    Output: path to inside archive
+    """
+    for file, folder in files:
+        torchaudio.datasets.utils.extract_archive(file)
+        path = os.path.dirname(file)
+        path = os.path.join(path, folder)
+        yield path
+
+
+def walk(paths, extension):
+    """
+    Walk inside a path recursively to find all files with given extension.
+
+    Input: path
+    Output: path, file name identifying a row of data
+    """
+    for path in paths:
+        for dp, dn, fn in os.walk(path):
+            for f in fn:
+                if extension in f:
+                    yield path, f
+
+
+def shuffle(generator):
+    """
+    Shuffle the order of a generator.
+
+    Input: generator
+    Output: generator
+    """
+
+    # Load whole generator in memory
+    generator = list(generator)
+    # print(len(generator))
+    random.shuffle(generator)
+    for g in generator:
+        yield g
+
+
+def filtering(fileids, reference):
+    """
+    Skip fileids that are not present in given reference file.
+
+    Output: (path, file) generator, reference file
+    Output: path, file
+    """
+
+    path_old = ""
+
+    for path, fileid in fileids:
+
+        # Check if same path to avoid reloading the file constantly
+        if path != path_old:
+            ref = os.path.join(path, reference)
+            with open(ref) as ref:
+                r = "".join(ref.readlines())
+            path_old = path
+
+        # It would be more efficient to loop through the reference file instead
+        if fileid in r:
+            yield path, fileid
+
+
+def load_yesno(fileids):
+    """
+    Load data corresponding to each YESNO fileids.
+
+    Input: path, file name identifying a row of data
+    Output: label, waveform, sample_rate
+    """
+
+    extension = ".wav"
+    for path, fileid in fileids:
+        file = os.path.join(path, fileid)
+        waveform, sample_rate = torchaudio.load(file)
+        label = os.path.basename(fileid).split(".")[0].split("_")
+
+        yield {"label": label, "waveform": waveform, "sample_rate": sample_rate}
+
+
+def YESNO(root):
+    """
+    Cache a pipeline loading YESNO.
```

Comment:
I'd say you're really building a pipeline here. The caching is optional and I'd expect that to see in a dataset class for convenience.

Reply:
The other thing we (I) semi under-appreciate is the fact that classes can print details about the pipeline pretty nicely. So that's also a nice side-effect of using a class for this.

Reply:
Agreed.